---
title: "Downsampled datasets"
output:
  workflowr::wflow_html:
    toc: true
    toc_depth: 2
editor_options:
  chunk_output_type: console
---

```{r setup, dev=c('png'), include=F}
knitr::opts_chunk$set(dev = c('png'))
knitr::opts_chunk$set(autodep = TRUE)

```



```{r include=F}
.libPaths("/net/snowwhite/home/aujackso/R/x86_64-pc-linux-gnu-library/4.0");
library(dplyr) 
library(readr)
library(ggplot2)
library(tibble)
library(edgeR)
library(DESeq2) 
library(RColorBrewer)
options(scipen=1, digits=5)
options(nwarnings=10000)

invnorm = function(x, seed) {
  set.seed(seed)
  qnorm((rank(x, na.last="keep", ties.method="random") - 0.5) / sum(!is.na(x)))
}


traits<-c("SEX")      
traits2<-c("SEX.M")   


```

RNA downsampled gene expression datasets supplied by Sarah.
NO checking for >=10 nuclei or genes 

```{r run_deseq_downsample, cache=FALSE, echo=FALSE, warning=FALSE, message=FALSE, include=TRUE, eval=FALSE}

traits<-c("SEX")      
traits2<-c("SEX.M")   

celltype<-c("Endothelial", "Macrophage", "Mesenchymal_Stem_Cell", "Neuromuscular_junction", "Neuronal", "Satellite_Cell", "Smooth_Muscle", "Type_2a", "Type_2x")
j<-1  #traits
  
pds.orig<-readRDS("data/pds.orig.Rds")



for (i in 1:length(celltype)) {
	eds.0<-read_tsv(paste0("/net/snowwhite/home/schanks/muscle/snrna/downsample/Type_1_",celltype[i],".downsample.tab"))
	genes<-eds.0$gene
	eds <- eds.0 %>% dplyr::select(-gene) %>% as.matrix()
	rownames(eds)<-genes
	print(dim(eds))

	who.use<-colnames(eds)

	pds.0 <- pds.orig[[10]] #Type_1, complete data
	print(dim(pds.0))
	pds <- pds.0 %>% dplyr::slice(match(who.use, labelcode))
	if (traits[j] %in% c("SEX","AGE")) {
	 	pds<-pds %>% 
		dplyr::select(SEX, labelcode, AGE, med.frac.mito, area_name, batch, tot_nuc_allcelltypes) %>% 
		dplyr::filter(complete.cases(.)) 
	}
	if (!(traits[j] %in% c("SEX","AGE"))) {
	 	pds<-pds %>% 
		dplyr::select(traits[j], SEX, labelcode, AGE, med.frac.mito, area_name, batch, tot_nuc_allcelltypes)  %>% 
		dplyr::filter(complete.cases(.)) 
	}
	pds<-pds %>% mutate(invmito=invnorm(med.frac.mito,1), invage=invnorm(AGE,1), invnucall=invnorm(tot_nuc_allcelltypes,1)) 	


	if (traits[j]=="SEX") {
		fstr<-"~ as.factor(batch) + as.factor(area_name) + invage + invmito + invnucall + as.factor(SEX)" 
		redstr<-"~ as.factor(batch) + as.factor(area_name) + invage + invmito + invnucall" 
	}
	if (traits[j]=="T2D") {
		fstr<-"~ as.factor(batch) + as.factor(area_name) + as.factor(SEX)+ invage + invmito + invnucall + T2D"
		redstr<-"~ as.factor(batch) + as.factor(area_name) + as.factor(SEX)+ invage + invmito + invnucall"
	}
	if (traits[j]=="AGE") {
		pds$x<- pds %>% pull(traits[j])
		pds$x<-invnorm(pds$x, j)
		fstr<-"~ as.factor(batch) + as.factor(area_name) +  as.factor(SEX) + invmito + invnucall + x"
		redstr<-"~ as.factor(batch) + as.factor(area_name) +  as.factor(SEX) + invmito + invnucall"
	}
	if (!(traits[j] %in% c("SEX", "AGE", "T2D"))) {
		pds$x<- pds %>% pull(traits[j])
		pds$x<-invnorm(pds$x, j)
		fstr<-"~ as.factor(batch) + as.factor(area_name) + as.factor(SEX) + invage + invmito + invnucall + x"
		redstr<-"~ as.factor(batch) + as.factor(area_name) + as.factor(SEX) + invage + invmito + invnucall"
	}
	dds <- DESeqDataSetFromMatrix(countData = eds,
            colData = pds, design = as.formula(fstr)) 
	dds<-scran::computeSumFactors(dds)   
	dds<-DESeq(dds, test="LRT",reduced=as.formula(redstr), useT=TRUE, minmu=1e-6, minReplicatesForReplace=Inf, fitType='local')  
	
	print(resultsNames(dds))
	stats<-mcols(dds) %>% as.data.frame() 
	stats$gene=rownames(stats)
	stats$cell=celltype[i]
	stats$trait=traits[j]
	converge<-stats %>% dplyr::select(gene, fullBetaConv)
	
	res = as(results(dds), "data.frame") 
	res$gene=rownames(res)
	res$cell=celltype[i]
	res$trait=traits2[j]
	res$n=nrow(pds)
	res<-merge(res,converge,by="gene")

map<-read_tsv("data/gencode.v30.gene_lengths_tss.tsv") %>% dplyr::select(gene_id, gene_name, chrom, gene_start, gene_end, gene_strand, gene_type)

res1<-merge(res, map, by.x="gene", by.y="gene_id", all.x=T) %>% arrange(pvalue)

		
write.table(format(res1, digits=8), 
file=paste0("output/DESeq.RNA/downsampled/Type_1_",celltype[i],".",traits2[j],".results.tab"), sep="\t", quote=F, col.names=T, row.names=F)


}

```

Could you please run Deseq on these downsampled ATAC-seq datasets (same as we did for RNA- without removing samples or peaks with few counts from these datasets)?  

/net/snowwhite/home/schanks/muscle/snatac/downsample/

Type_1_Adipocyte.downsample.tab
Type_1_Endothelial.downsample.tab
Type_1_Macrophage.downsample.tab
Type_1_Mesenchymal_Stem_Cell.downsample.tab
Type_1_Neuromuscular_junction.downsample.tab
Type_1_Neuronal.downsample.tab
Type_1_Satellite_Cell.downsample.tab
Type_1_Smooth_Muscle.downsample.tab
Type_1_T_cell.downsample.tab
Type_1_Type_2a.downsample.tab
Type_1_Type_2x.downsample.tab

```{r run_deseq_downsample_atac, cache=FALSE, echo=FALSE, warning=FALSE, message=FALSE, include=TRUE, eval=FALSE}

traits<-c("SEX")      
traits2<-c("SEX.M")   

celltype<-c("Adipocyte", "Endothelial", "Macrophage", "Mesenchymal_Stem_Cell", "Neuromuscular_junction", "Neuronal", "Satellite_Cell", "Smooth_Muscle", "T_cell", "Type_2a", "Type_2x")
j<-1  #traits
  
pds.orig<-readRDS(paste0("~/sn_muscle_2023/data/ATAC.pds.orig.Type_1.Rds")) #280 samples



for (i in 1:length(celltype)) {
#/net/snowwhite/home/schanks/muscle/snrna/downsample/genefix/*.tab
	eds.0<-read_tsv(paste0("/net/snowwhite/home/schanks/muscle/snatac/downsample/Type_1_",celltype[i],".downsample.tab"))
	peaks<-eds.0$peak
	eds <- eds.0 %>% dplyr::select(-peak) %>% as.matrix()
	rownames(eds)<-peaks
	print(dim(eds))

	who.use<-colnames(eds)

	pds.0 <- pds.orig #Type_1
	print(dim(pds.0))
	pds <- pds.0 %>% dplyr::slice(match(who.use, labelcode))
	print(dim(pds))
	
	if (traits[j] %in% c("SEX","AGE")) {
	 	pds<-pds %>% 
		dplyr::select(SEX, labelcode, AGE, area_name, tss_enrichment, batch, fraction_mitochondrial_median, tot_nuc_allcelltypes) %>% 
		filter(complete.cases(.)) 
	}
	
	if (!(traits[j] %in% c("SEX","AGE"))) {
	 	pds<-pds %>% 
		dplyr::select(traits[j], SEX, labelcode, AGE, area_name, tss_enrichment, batch, fraction_mitochondrial_median, tot_nuc_allcelltypes) %>% 
		filter(complete.cases(.)) 
	}
	
	pds<-pds %>% 
		mutate(invtssenrich=invnorm(tss_enrichment,1), invage=invnorm(AGE,1), 
		invmito=invnorm(fraction_mitochondrial_median,1), 
		invtotnuc=invnorm(tot_nuc_allcelltypes, 1))  %>%
		mutate(batch2=gsub("-NM","",batch)) %>% 
		mutate(batch3=paste0("B",as.character(batch2))) %>%
		dplyr::select(-batch, -batch2) %>% 
		dplyr::rename(batch=batch3)	



	if (traits[j]=="SEX") {
		fstr<-"~ as.factor(batch) + as.factor(area_name) + invage + invtssenrich + invmito + invtotnuc + as.factor(SEX)" 
		redstr<-"~ as.factor(batch) + as.factor(area_name) + invage + invtssenrich + invmito + invtotnuc" 
	}
	if (traits[j]=="T2D") {
		fstr<-"~ as.factor(batch) + as.factor(area_name) + as.factor(SEX) + invage + invtssenrich + invmito + invtotnuc + T2D"
		redstr<-"~ as.factor(batch) + as.factor(area_name) + as.factor(SEX) + invage + invtssenrich + invmito + invtotnuc"
	}
	if (traits[j]=="AGE") {
		pds$x<- pds %>% pull(traits[j])
		pds$x<-invnorm(pds$x, j)
		fstr<-"~ as.factor(batch) + as.factor(area_name) + as.factor(SEX) + invtssenrich + invmito + invtotnuc + x"
		redstr<-"~ as.factor(batch) + as.factor(area_name) + as.factor(SEX) + invtssenrich + invmito + invtotnuc"
	}
	if (!(traits[j] %in% c("SEX", "AGE", "T2D"))) {
		pds$x<- pds %>% pull(traits[j])
		pds$x<-invnorm(pds$x, j)
		fstr<-"~ as.factor(batch) + as.factor(area_name) + as.factor(SEX) + invage + invtssenrich + invmito + invtotnuc + x"
		redstr<-"~ as.factor(batch) + as.factor(area_name) + as.factor(SEX) + invage + invtssenrich + invmito + invtotnuc"
	}
	
	# beware count 0 samples
	totct<-colSums(eds) %>% as.data.frame()
	colnames(totct)<-"totct"
	totct<-tibble::rownames_to_column(totct,'labelcode')
	who.use<-totct %>% filter(totct>0) %>% pull(labelcode)
	eds.use<-eds %>% as.data.frame() %>% dplyr::select(all_of(who.use)) %>% as.matrix()
	pds.use<-pds %>% as.data.frame() %>% dplyr::slice(match(who.use, labelcode))
	dds <- DESeqDataSetFromMatrix(countData = eds.use,
            colData = pds.use, design = as.formula(fstr)) 
	dds<-scran::computeSumFactors(dds)   
	dds<-DESeq(dds, test="LRT",reduced=as.formula(redstr), useT=TRUE, minmu=1e-6, minReplicatesForReplace=Inf, fitType='local')  
	
	print(resultsNames(dds))
	stats<-mcols(dds) %>% as.data.frame() 
	stats$peak=rownames(stats)
	stats$cell=celltype[i]
	stats$trait=traits[j]
	converge<-stats %>% dplyr::select(peak, fullBetaConv)
	
	res = as(results(dds), "data.frame") 
	res$peak=rownames(res)
	res$cell=celltype[i]
	res$trait=traits2[j]
	res$n=nrow(pds)
	res<-merge(res,converge,by="peak")

res1<-res %>% arrange(pvalue)

		
write.table(format(res1, digits=8), 
file=paste0("output/DESeq.ATAC/downsampled/Type_1_",celltype[i],".",traits2[j],".results.tab"), sep="\t", quote=F, col.names=T, row.names=F)


}

```


